src/main/scala/org/ergoplatform/http/api/WalletApiRoute.scala

```diff
@@ -122,32 +125,85 @@ case class WalletApiRoute(readersHolder: ActorRef, nodeViewActorRef: ActorRef, e
-      case Success(tx) => ApiResponse(tx)
+      case Success(tx) => processFn(tx)
     }
   }
 
-  private def sendTransaction(requests: Seq[TransactionRequest], inputsRaw: Seq[String]): Route = {
-    withWalletOp(_.generateTransaction(requests, inputsRaw)) {
-      case Failure(e) =>
-        BadRequest(s"Bad request $requests. ${Option(e.getMessage).getOrElse(e.toString)}")
-      case Success(tx) =>
-        nodeViewActorRef ! LocallyGeneratedTransaction[ErgoTransaction](tx)
-        ApiResponse(tx.id)
-    }
+  private def generateTransaction(requests: Seq[TransactionGenerationRequest],
+                                  inputsRaw: Seq[String],
+                                  dataInputsRaw: Seq[String]): Route = {
+    generateTransactionAndProcess(requests, inputsRaw, dataInputsRaw, tx => ApiResponse(tx))
+  }
+
+  private def sendTransaction(requests: Seq[TransactionGenerationRequest],
+                              inputsRaw: Seq[String],
+                              dataInputsRaw: Seq[String]): Route = {
+    generateTransactionAndProcess(requests, inputsRaw, dataInputsRaw, {tx =>
+      nodeViewActorRef ! LocallyGeneratedTransaction[ErgoTransaction](tx)
+      ApiResponse(tx.id)
+    })
   }
 
-  def sendTransactionR: Route = (path("transaction" / "send") & post
-    & entity(as[RequestsHolder])) (holder => sendTransaction(holder.withFee, holder.inputsRaw))
+  def sendTransactionR: Route =
+    (path("transaction" / "send") & post & entity(as[RequestsHolder])){ holder =>
+      sendTransaction(holder.withFee, holder.inputsRaw, holder.dataInputsRaw)
+    }
+
+  def generateTransactionR: Route =
+    (path("transaction" / "generate") & post & entity(as[RequestsHolder])){ holder =>
+      generateTransaction(holder.withFee, holder.inputsRaw, holder.dataInputsRaw)
+    }
+
+  def signTransactionR: Route = (path("transaction" / "sign")
+    & post & entity(as[TransactionSigningRequest])) { tsr =>
+
+    val tx = tsr.utx
+    val secrets = (tsr.dlogs ++ tsr.dhts).map(ExternalSecret.apply)
+
+    def signWithReaders(r: Readers): Future[Try[ErgoTransaction]] = {
+      if (tsr.inputs.isDefined) {
+        val boxesToSpend = tsr.inputs.get
+          .flatMap(in => Base16.decode(in).flatMap(ErgoBoxSerializer.parseBytesTry).toOption)
+        val dataBoxes = tsr.dataInputs.getOrElse(Seq.empty)
+          .flatMap(in => Base16.decode(in).flatMap(ErgoBoxSerializer.parseBytesTry).toOption)
 
-  def generateTransactionR: Route = (path("transaction" / "generate") & post
-    & entity(as[RequestsHolder])) (holder => generateTransaction(holder.withFee, holder.inputsRaw))
+        if (boxesToSpend.size == tx.inputs.size && dataBoxes.size == tx.dataInputs.size) {
+          r.w.signTransaction(secrets, tx, boxesToSpend, dataBoxes)
+        } else {
+          Future(Failure(new Exception("Can't parse input boxes provided")))
+        }
+      } else {
+        r.s match {
+          case utxoSet: UtxoStateReader =>
```

Comment:
Would be nice to fetch boxes from the mempool here too, so that chained transactions can be easily signed.

Reply:
Looks like this is done now. Thanks!

Reply:
Yes, done!